pkg/machines/components/vcpuModal.jsx

```diff
@@ -0,0 +1,241 @@
+import React from 'react';
+import cockpit from 'cockpit';
+
+import { show_modal_dialog } from 'cockpit-components-dialog.jsx';
+import SelectComponent from 'cockpit-components-select.jsx';
+import InfoRecord from './infoRecord.jsx';
+import { setVCPUSettings } from "../actions.es6";
+
+React;
+
+const _ = cockpit.gettext;
+
+const dividers = (num) => {
+    const divs = [1];
+
+    for (let i = 2; i < num; i++) {
+        if (num % i === 0) {
+            divs.push(i);
+        }
+    }
+
+    if (num > 1) {
+        divs.push(num);
+    }
+
+    return divs;
+}
+
+const clamp = (value, max) => {
+    return value > max ? max : value;
+}
+
+const Select = function ({ id, items, onChange, value }) {
+    return ( <SelectComponent.Select id={id} initial={value} onChange={onChange}>
+        {items.map((t) => (
+            <SelectComponent.SelectEntry data={t}>{t}</SelectComponent.SelectEntry>
+        ))}
+    </SelectComponent.Select>);
+}
+
+class VCPUModalBody extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            max: props.vcpus.max || 1,
+            count: parseInt(props.vcpus.count) || 1,
+            sockets: props.cpu.topology.sockets || 1,
+            threads: props.cpu.topology.threads || 1,
+            cores: props.cpu.topology.cores || 1
+        };
+        this.onMaxChange = this.onMaxChange.bind(this);
+        this.onCountSelect = this.onCountSelect.bind(this);
+        this.onSocketChange = this.onSocketChange.bind(this);
+        this.onThreadsChange = this.onThreadsChange.bind(this);
+        this.onCoresChange = this.onCoresChange.bind(this);
+        props.onChange(this.state);
+    }
+
+    componentWillUpdate (nextProps, nextState) {
+        this.props.onChange(nextState);
+    }
+
+    onMaxChange (e) {
+        const maxHypervisor = parseInt(this.props.hypervisorMax);
+        let maxValue = parseInt(e.target.value);
+
+        // Check new value for limits
+        if (isNaN(maxValue) || maxValue <= 0) {
+            maxValue = 1;
+        }
+        maxValue = clamp(maxValue, maxHypervisor);
+
+        // If count of used VCPU greater then new max value, then change it to new max value
+        if (maxValue < this.state.count) {
+            this.setState({ count: maxValue });
+        }
+
+        // Recalculate new values for sockets, cores and threads according to new max value
+        // Max value = Sockets * Cores * Threads
+        let state = { max: maxValue, sockets: this.state.sockets, cores: this.state.cores };
+
+        // Recalculate sockets first, and get array of all divisors of new max values
+        let divs = dividers(state.max);
+
+        // If current sockets value is not in divisors array, then change it to max divisor
+        if (divs.indexOf(this.state.sockets) === -1 || (this.props.cpu.topology.sockets || 1) === this.state.sockets) {
+            state.sockets = divs[divs.length - 1];
+        }
+
+        // Get next divisors 
```

Comment:
This has trailing space, in some other comments too. Can you please search for ` $` and drop them? git yells rather loudly about these.